src/aro/Value.zig

```diff
@@ -390,9 +520,65 @@ pub fn sub(res: *Value, lhs: Value, rhs: Value, ty: Type, comp: *Compilation) !b
     }
 }
 
+/// c99 Annex G complex multiplication algorithm
+pub fn complexMul(lhs: Value, rhs: Value, comptime T: type, comp: *Compilation) !Value {
+    var a = lhs.real(T, comp);
+    var b = lhs.imag(T, comp);
+    var c = rhs.real(T, comp);
+    var d = rhs.imag(T, comp);
+    const ac = a * c;
+    const bd = b * d;
+    const ad = a * d;
+    const bc = b * c;
+    var x = ac - bd;
+    var y = ad + bc;
+    if (std.math.isNan(x) and std.math.isNan(y)) {
+        var recalc = false;
+        if (std.math.isInf(a) or std.math.isInf(b)) {
+            // lhs infinite
+            // Box the infinity and change NaNs in the other factor to 0
+            a = std.math.copysign(if (std.math.isInf(a)) @as(T, 1.0) else @as(T, 0.0), a);
+            b = std.math.copysign(if (std.math.isInf(b)) @as(T, 1.0) else @as(T, 0.0), b);
+            if (std.math.isNan(c)) c = std.math.copysign(@as(T, 0.0), c);
+            if (std.math.isNan(d)) d = std.math.copysign(@as(T, 0.0), d);
+            recalc = true;
+        }
+        if (std.math.isInf(c) or std.math.isInf(d)) {
+            // rhs infinite
+            // Box the infinity and change NaNs in the other factor to 0
+            c = std.math.copysign(if (std.math.isInf(c)) @as(T, 1.0) else @as(T, 0.0), c);
+            d = std.math.copysign(if (std.math.isInf(d)) @as(T, 1.0) else @as(T, 0.0), d);
+            if (std.math.isNan(a)) a = std.math.copysign(@as(T, 0.0), a);
+            if (std.math.isNan(b)) b = std.math.copysign(@as(T, 0.0), b);
+            recalc = true;
+        }
+        if (!recalc and (std.math.isInf(ac) or std.math.isInf(bd) or std.math.isInf(ad) or std.math.isInf(bc))) {
+            // Recover infinities from overflow by changing NaN's to 0
+            if (std.math.isNan(a)) a = std.math.copysign(@as(T, 0.0), a);
+            if (std.math.isNan(b)) b = std.math.copysign(@as(T, 0.0), b);
+            if (std.math.isNan(c)) c = std.math.copysign(@as(T, 0.0), c);
+            if (std.math.isNan(d)) d = std.math.copysign(@as(T, 0.0), d);
+        }
+        if (recalc) {
+            x = std.math.inf(T) * (a * c - b * d);
+            y = std.math.inf(T) * (a * d + b * c);
+        }
+    }
+    return makeComplex(T, x, y, comp);
+}
+
 pub fn mul(res: *Value, lhs: Value, rhs: Value, ty: Type, comp: *Compilation) !bool {
     const bits: usize = @intCast(ty.bitSizeof(comp).?);
-    if (ty.isFloat()) {
+    if (ty.isComplex()) {
+        res.* = switch (bits) {
+            64 => try lhs.complexMul(rhs, f32, comp),
+            128 => try lhs.complexMul(rhs, f64, comp),
+            160 => try lhs.complexMul(rhs, f80, comp),
+            256 => try lhs.complexMul(rhs, f128, comp),
+            else => unreachable,
+        };
```

Comment:
For future int support it'd be nice if this `isComplex` check was moved to the `isFloat` branch which then called `complexMulFloat`.